omp/solver/gmres_kernels.cpp

```diff
@@ -48,6 +49,200 @@ namespace omp {
 namespace gmres {
 
 
+namespace {
+
+
+template <typename ValueType>
+void finish_arnoldi(matrix::Dense<ValueType> *next_krylov_basis,
+                    matrix::Dense<ValueType> *krylov_bases,
+                    matrix::Dense<ValueType> *hessenberg_iter, size_type iter,
+                    const stopping_status *stop_status)
+{
+#pragma omp declare reduction(add : ValueType : omp_out = omp_out + omp_in)
+
+    for (size_type i = 0; i < next_krylov_basis->get_size()[1]; ++i) {
+        if (stop_status[i].has_stopped()) {
+            continue;
+        }
+        for (size_type k = 0; k < iter + 1; ++k) {
+            ValueType hessenberg_iter_entry = 0;
+
+#pragma omp parallel for reduction(add : hessenberg_iter_entry)
```

Comment:
I could imagine that this parallelization actually slows down - as j is usually of moderate size, and the parallelization requires a global reduction.

Reply:
Currently, this implementation is slower in most cases than the reference implementation. We might have to go back to his simple version where make the outer loop parallel (which is over the number of right hand sides).
I will see if I can make it faster with the current approach.

Reply:
I don't necessarily agree. I think we instead should try without the (reduction-requiring) parallelization of operations across the krylov subspace. Parallelizing across the rows (much higher parallelism than the krylow subspace) should still give benefits.

Reply:
As far as I can tell, he already did that. The only `for`-loop that is further out is the loop over the number of right hand side. I will simply try it with the removed reduction loops and see how it performs.

Reply:
@thoasm do you have any performance numbers?

Reply:
I modified the tests in order to get the time of each kernel. Additionally, I set `m = 1597` (matrix width and height) and `n = 43` (number of RHS).

I ran all tests on my laptop with 4 threads, each kernel ran 10x, then averaged.
The CI system is very busy at the moment, so we might get invalid results there.
```
Current version
Initialize1:
    Reference iteration: 310948 ns
    OpenMP iteration:    231908 ns
Initialize2:
    Reference iteration: 19551071 ns
    OpenMP iteration:    10425265 ns
Step1:
    Reference iteration: 5331477 ns
    OpenMP iteration:    4568243 ns
Step2:
    Reference iteration: 1153782 ns
    OpenMP iteration:    422260 ns


Without reduce for-loops
Initialize1:
    Reference iteration: 319508 ns
    OpenMP iteration:    240941 ns
Initialize2:
    Reference iteration: 19229232 ns
    OpenMP iteration:    11102717 ns
Step1:
    Reference iteration: 6080392 ns
    OpenMP iteration:    5206029 ns
Step2:
    Reference iteration: 1181854 ns
    OpenMP iteration:    363137 ns
    
Parallel over number RHS
Initialize1:
    Reference iteration: 313300 ns
    OpenMP iteration:    322662 ns
Initialize1:
    Reference iteration: 21236060 ns
    OpenMP iteration:    21861582 ns
Step1:
    Reference iteration: 5351436 ns
    OpenMP iteration:    2060129 ns
Step2:
    Reference iteration: 1128927 ns
    OpenMP iteration:    414221 ns
```

The current version actually looks very good for large `m`, but for smaller `m`, OpenMP is slower in some cases.

Reply:
I will now also test different `schedule`s with the bigger matrix, previously (with `m = 597`), it got worse.